sdk/src/main/java/com/uid2/utils/InputUtils.kt

```diff
@@ -0,0 +1,163 @@
+import com.uid2.utils.InputUtils.EmailParsingState.SubDomain
+import java.util.Locale
+
+/**
+ * This class contains a set of methods that will attempt to normalize a given [IdentityRequest]. This is to validate
+ * what is provided to us to ensure it's in the expected format, before we attempt to consume it.
+ */
+internal class InputUtils {
+
+    /**
+     * Attempts to normalize the given [IdentityRequest.Phone]
+     *
+     * @throws InputValidationException Thrown if the input could not be normalized.
+     */
+    @Throws(InputValidationException::class)
+    fun normalize(request: IdentityRequest.Phone): IdentityRequest.Phone {
+        if (!isNormalized(request)) {
+            throw InputValidationException("Phone number is not normalized to ITU E.164 standard")
+        }
+
+        return request
+    }
+
+    /**
+     * Attempts to normalize the given [IdentityRequest.Email]
+     *
+     * @throws InputValidationException Thrown if the input could not be normalized.
+     */
+    @Throws(InputValidationException::class)
+    fun normalize(request: IdentityRequest.Email): IdentityRequest.Email {
+        val normalized = normalizeEmail(request.data)
+            ?: throw InputValidationException("Invalid email address detected")
+
+        return IdentityRequest.Email(normalized)
+    }
+
+    /**
+     * Returns whether or not the given [IdentityRequest.Phone] has already been normalized. This will check against
+     * the ITU E.164 Standard (https://en.wikipedia.org/wiki/E.164).
+     */
+    private fun isNormalized(request: IdentityRequest.Phone): Boolean {
+        val number = request.data
+
+        // Firstly, let's check to make sure we have a non-empty string
+        if (number.isEmpty()) {
+            return false
+        }
+
+        // The first character should be a '+'.
+        if (number[0] != '+') {
+            return false
+        }
+
+        // Check to make sure that only digits are contained.
+        val allDigits = number.substring(1).all { isAsciiDigit(it) }
+        if (!allDigits) {
+            return false
+        }
+
+        // The number of digits (excluding the '+') should be in the expected range.
+        val totalDigits = number.length - 1
+        return !(totalDigits < MIN_PHONE_NUMBER_DIGITS || totalDigits> MAX_PHONE_NUMBER_DIGITS)
+    }
+
+    /**
+     * Returns whether or not the given [String] contains only ASCII digits.
+     */
+    private fun isAsciiDigit(digit: Char): Boolean {
+        return digit in '0'..'9'
+    }
+
+    private enum class EmailParsingState {
+        Starting,
+        SubDomain,
+    }
+
+    /**
+     * This code will attempt to normalize a given email address. It's been translated from a Java reference, and
+     * therefore been kept as close to the original implementation as possible (with some unused conditions removed).
+     *
+     * https://github.com/IABTechLab/uid2-operator/blob/main/src/main/java/com/uid2/operator/service/InputUtil.java#L96
```

Comment:
nit: This line reference is pointing to the main commit, but it is susceptible to change in the repo and it won't point to what you intend then. You might point to a specific commit so the line will always point to what you want.

Reply:
Sounds good, thanks